### ISSUE TITLE
Cannot open the preferences panel either from ibus-setup or GNOME-settings

### ISSUE THREAD

**Issue by tzihad** [sarim/ibus-avro#179]
The error message :
``` Gjs-Message: 14:21:23.901: JS WARNING: [/usr/share/ibus-avro/pref.js 30]: Requiring Gtk but it has 3 versions available; use imports.gi.versions to pick one         ─╯
Gjs-Message: 14:21:23.914: JS WARNING: [/usr/share/ibus-avro/pref.js 38]: Too many arguments to function Gtk.init: expected 0, got 2

(gjs:4544): Gjs-CRITICAL **: 14:21:23.918: JS ERROR: Gtk.BuilderError: /usr/share/ibus-avro/avropref.ui:15:28 Invalid property: GtkWindow.icon
runpref@/usr/share/ibus-avro/pref.js:40:13
@/usr/share/ibus-avro/pref.js:90:5


(gjs:4544): Gjs-CRITICAL **: 14:21:23.918: Script /usr/share/ibus-avro/pref.js threw an exception
```
I am using the ibus-avro-git from aur.

**Comment by sarim** [comment]
let us know the versions of relevant libraries in your system. Gjs, Gtk, Gnome.

**Comment by tzihad** [comment]
gjs 1.68.1
gtk 4.2.1
gnome 40.2.0

**Comment by gunnarhj** [comment]
FWIW I can not reproduce the issue on Ubuntu 21.10, where `gjs 1.68.1` is present but gtk is still gtk3 (3.24.29). "gnome" is ambiguous — Ubuntu 21.10 currently has a mix of 3.38 and 40.

@tzihad: Since you seem to have gtk4, is `ibus-gtk4` installed?

**Comment by sarim** [comment]
>> (gjs:4544): Gjs-CRITICAL **: 14:21:23.918: JS ERROR: Gtk.BuilderError: /usr/share/ibus-avro/avropref.ui:15:28 Invalid property: GtkWindow.icon

It is because of libgtk4. In the [gtk4 documentation of GtkWindow](https://developer.gnome.org/gtk4/stable/GtkWindow.html), I don't see "icon" properties anymore, seems like it was removed. Its there for both gtk2 and gtk3 :P

I'll explicitly set Gtk version to 3 in preferences.

**Comment by tzihad** [comment]
Okay, thanks for the response.

**Comment by sarim** [comment]
@tzihad What is your OS and version? Can you point me to a live iso for testing?

**Comment by tzihad** [comment]
I use Arch Linux, All packages are synced with xeonbd's mirror,

**Comment by sarim** [comment]
@tzihad I just pushed a commit (2883b96) fixing the issue, If you update your system's `/usr/share/ibus-avro/pref.js` with the file from this repository the problem will be fixed. Or inform the maintainer the of package to update.

@gunnarhj Same problem also occurs in Ubuntu 21.04 if both gtk3 and gtk4 is available to gjs. After booting the live cd, I installed `sudo apt install libgtk-4-1 gir1.2-gtk-4.0`. After this gjs imports gtk4 if the application doesn't specify version. 
![image](https://user-images.githubusercontent.com/1235888/124392013-39e08800-dd15-11eb-834f-1839cfdb1bf2.png)

Funny thing Ubiquity also just crashed and from the error log it looks like same issue :P 

However, the point is `ibus-avro` package in ubuntu 21.04 also needs to be updated with fix I just pushed. Can you get the ball rolling on that? Should I published another minor version?

**Comment by tzihad** [comment]
Thanks! It works now.

**Comment by gunnarhj** [comment]
> Same problem also occurs in Ubuntu 21.04 if both gtk3 and gtk4 is available to gjs. After booting the live cd, I installed `sudo apt install libgtk-4-1 gir1.2-gtk-4.0`.

Hmm.. I thought that gtk4 didn't exist in the Ubuntu 21.04 archive, but now I see that it was copied from Debian experimental before the release.

> `ibus-avro` package in ubuntu 21.04 also needs to be updated with fix I just pushed. Can you get the ball rolling on that?

Sure. But I'd better start with Debian and 21.10 to follow the procedures.

> Should I published another minor version?

Not necessary — I can patch it as part of the packaging.

Btw, 276fcafa is still there, but last time I asked, you didn't consider it ready for Debian/Ubuntu. I'm assuming that's still the case.

**Comment by sarim** [comment]
> Btw, [276fcaf](https://github.com/sarim/ibus-avro/commit/276fcafad9b562ef32e88a2dabdb4dfeee1b2d72) is still there, but last time I asked, you didn't consider it ready for Debian/Ubuntu. I'm assuming that's still the case.

Hmm, not sure what to do with that. That commit reorganizes stuff around a bit that may or may not help with GC. And after further research It didn't provide any substantial benefit to be problem (If any). So it doesn't provide any real value, It doesn't break anything. So I guess there is no need to add that as a patch.

**Comment by tzihad** [comment]
@sarim There are some other issues,Not sure I should report is as an another issue or not.Anyways, ibus-avro always report these errors (running with -vx, and dictionary suggestion on in preferences) after typing for a while. 
```
(gjs:95132): Gjs-CRITICAL **: 00:44:41.991: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.

(gjs:95132): Gjs-CRITICAL **: 00:44:41.991: The offending signal was process-key-event on IBusEngine 0x55928a4f2980.
== Stack trace for context 0x5592898f2140 ==
#0   5592899184f0 i   /usr/share/ibus-avro/main-gjs.js:433 (1d918429b3d0 @ 1850)
```
Not sure what triggers this error.

**Comment by sarim** [comment]
@tzihad Thats #156 . It is a bug of gjs/gnome. There is no easy solution for it ATM.

**Comment by tzihad** [comment]
Thanks, again

**Comment by gunnarhj** [comment]
@sarim: FYI the fix of Ubuntu 21.04 is [on its way](https://launchpad.net/bugs/1934606). May take a couple of weeks.

**Comment by sarim** [comment]
@gunnarhj That was fast. You're awesome as always :3 Thanks :)
